source/microbit/modevents.cpp

```diff
@@ -0,0 +1,582 @@
+
+STATIC mp_obj_t events_tick_once(mp_obj_t ms_in) {
+    mp_int_t delay_ms = mp_obj_get_int(ms_in);
+    return mp_obj_new_int(add_tick_once_scanner(delay_ms));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(events_tick_once_obj, events_tick_once);
+
+STATIC mp_obj_t events_when_button_a_pressed(void) {
+    return mp_obj_new_int(add_button_scanner(MICROBIT_ID_BUTTON_A));
+}
+MP_DEFINE_CONST_FUN_OBJ_0(events_when_button_a_pressed_obj, events_when_button_a_pressed);
+
+STATIC mp_obj_t events_when_button_b_pressed(void) {
+    return mp_obj_new_int(add_button_scanner(MICROBIT_ID_BUTTON_B));
+}
+MP_DEFINE_CONST_FUN_OBJ_0(events_when_button_b_pressed_obj, events_when_button_b_pressed);
+
+STATIC mp_obj_t events_when_compass_between(mp_obj_t angle1_in, mp_obj_t angle2_in) {
+    mp_int_t angle1 = mp_obj_get_int(angle1_in);
+    mp_int_t angle2 = mp_obj_get_int(angle2_in);
+    return mp_obj_new_int(add_compass_scanner(angle1, angle2));
+}
+MP_DEFINE_CONST_FUN_OBJ_2(events_when_compass_between_obj, events_when_compass_between);
+
+STATIC mp_obj_t events_when_accelerometer_x_between(mp_obj_t v1_in, mp_obj_t v2_in) {
+    mp_int_t v1 = mp_obj_get_int(v1_in);
+    mp_int_t v2 = mp_obj_get_int(v2_in);
+    return mp_obj_new_int(add_accelerometer_scanner(MICROBIT_EVENTS_ACCELEROMETER_X, v1, v2));
+}
+MP_DEFINE_CONST_FUN_OBJ_2(events_when_accelerometer_x_between_obj, events_when_accelerometer_x_between);
+
+STATIC mp_obj_t events_when_accelerometer_y_between(mp_obj_t v1_in, mp_obj_t v2_in) {
+    mp_int_t v1 = mp_obj_get_int(v1_in);
+    mp_int_t v2 = mp_obj_get_int(v2_in);
+    return mp_obj_new_int(add_accelerometer_scanner(MICROBIT_EVENTS_ACCELEROMETER_Y, v1, v2));
+}
+MP_DEFINE_CONST_FUN_OBJ_2(events_when_accelerometer_y_between_obj, events_when_accelerometer_y_between);
+
+STATIC mp_obj_t events_when_accelerometer_z_between(mp_obj_t v1_in, mp_obj_t v2_in) {
+    mp_int_t v1 = mp_obj_get_int(v1_in);
+    mp_int_t v2 = mp_obj_get_int(v2_in);
+    return mp_obj_new_int(add_accelerometer_scanner(MICROBIT_EVENTS_ACCELEROMETER_Z, v1, v2));
+}
+MP_DEFINE_CONST_FUN_OBJ_2(events_when_accelerometer_z_between_obj, events_when_accelerometer_z_between);
+
+mp_obj_t microbit_events(void){
+    events_t *result = m_new_obj(events_t);
+    result->base.type = &microbit_events_type;
+    return result;
+}
+
+STATIC mp_obj_t get_microbit_events_iter(mp_obj_t o_in) {
+    events_t *events = (events_t *)o_in;
+    events_iterator_t *result = m_new_obj(events_iterator_t);
+    result->base.type = &microbit_events_iterator_type;
+    result->events = events;
+    return result;
+}
+
+STATIC mp_obj_t microbit_events_iter_next(mp_obj_t o_in) {
+    // Is there a way to mark o_in as unused?
+    uint16_t event_id;
+    microbit_events_obj->state = MICROBIT_EVENTS_LOOP_RUNNING;
+    event_id = event_queue_dequeue(microbit_events_obj->event_queue);
+    if (event_id == MICROBIT_EVENTS_NO_EVENT) {
+        // Is there a way to indicate that the iterator should not yield?
```

Comment:
`__WFI();` or `__WFE();`, it depends on interrupt level. Probably `__WFE();` in this case.